lib/src/repository_url.dart

```diff
@@ -0,0 +1,180 @@
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:path/path.dart' as p;
+
+/// Describes the url + folder or file path of a repository URL.
+class RepositoryUrl {
+  /// One of the values from [RepositoryProvider].
+  final String provider;
+
+  /// The base URL up to the repository itself.
+  final String baseUrl;
+
+  /// The separator between the repository and the relative folder or file path.
+  final String? separator;
+
+  /// The branch to use.
+  final String? branch;
+
+  /// The relative folder or path of the repository.
+  final String path;
+
+  RepositoryUrl({
+    this.provider = RepositoryProvider.unknown,
+    required this.baseUrl,
+    this.separator,
+    this.branch,
+    required this.path,
+  });
+
+  /// Parses [input] and return the parsed [RepositoryUrl] if successful,
+  /// or returns null if it was unable to recognize the pattern.
+  static RepositoryUrl? tryParse(String input) => _tryParseRepositoryUrl(input);
+
+  /// Parses [input] and return the parsed [RepositoryUrl] if successful,
+  /// or throws [FormatException] if it was unable to recognize the pattern.
+  static RepositoryUrl parse(String input) {
+    final v = tryParse(input);
+    if (v == null) {
+      throw FormatException('Unable to parse repository url: `$input`.');
+    } else {
+      return v;
+    }
+  }
+
+  /// Resolves a relative path and returns a new instance of the [RepositoryUrl].
+  RepositoryUrl resolve(
+    String relativePath, {
+    String? separator,
+    String? branch,
+  }) {
+    final newPath = p.joinAll([
+      if (path.isNotEmpty) path,
+      if (relativePath.isNotEmpty) relativePath,
+    ]);
+    final normalizedNewPath = p.normalize(newPath);
+
+    if (RepositoryProvider.isGitHubCompatible(provider)) {
+      final extension = p.extension(normalizedNewPath).toLowerCase();
+      final needsRaw = _imageExtensions.contains(extension);
+      final newBaseUrl = baseUrl.endsWith('.git')
+          ? baseUrl.substring(0, baseUrl.length - 4)
+          : baseUrl;
+      return RepositoryUrl(
+        provider: provider,
+        baseUrl: newBaseUrl,
+        separator: separator ?? (needsRaw ? 'raw' : null) ?? 'blob',
+        branch: branch ?? this.branch ?? 'master',
+        path: normalizedNewPath,
+      );
+    } else {
+      return RepositoryUrl(
+        provider: provider,
+        baseUrl: baseUrl,
+        separator: separator,
+        branch: branch,
+        path: normalizedNewPath,
+      );
+    }
+  }
+
+  /// Creates a String representation of the URL.
+  String toUrl() =>
+      p.joinAll([baseUrl, separator, branch, path].whereType<String>());
+}
+
+abstract class RepositoryProvider {
+  static const github = 'github';
+  static const gitlab = 'gitlab';
+  static const unknown = 'unknown';
+
+  static bool isGitHubCompatible(String provider) =>
+      provider == github || provider == gitlab;
+}
+
+const _repoReplacePrefixes = {
+  'http://github.com': 'https://github.com',
```

Comment:
Perhaps this replacement is better done by changing the scheme of the parsed url.

```
url = url.replace(scheme:'https')
``

Reply:
Done.